tools/castle-editor/code/castleundosystem.pas

```diff
@@ -0,0 +1,276 @@
+    function Redo: TUndoHistoryElement;
+    { If it's possible to perform Undo right now }
+    function IsUndoPossible: Boolean;
+    { If it's possible to perform Redo right now }
+    function IsRedoPossible: Boolean;
+    { Get comment for Undo action if it's possible, or return '' }
+    function UndoComment: String;
+    { Get comment for Redo action if it's possible, or return '' }
+    function RedoComment: String;
+    { Clear all Undo History }
+    procedure ClearUndoHistory;
+  end;
+
+implementation
+uses
+  SysUtils,
+  CastleComponentSerialize, CastleLog;
+
+function TUndoHistoryElement.Size: SizeInt;
+
+  //see https://www.freepascal.org/docs-html/ref/refsu13.html
+  //I'm not sure if this includes size of string pointer or not
+  //Note that Length(AnsiString) returns length of the string in bytes, not in UTF8 characters
+  function SizeOfAnsiString(const S: String): SizeInt;
+  const
+    HS = 16;
+  begin
+    Result := Length(S) + 1 + HS;
+  end;
+
+begin
+  Result := Self.InstanceSize + SizeOfAnsiString(Data) + SizeOfAnsiString(Selected) + SizeOfAnsiString(Comment);
+end;
+
+constructor TUndoSystem.Create(AOwner: TComponent);
+begin
+  inherited;
+  UndoHistory := TUndoHistory.Create(true);
+  ClearUndoHistory;
+end;
+
+destructor TUndoSystem.Destroy;
+begin
+  FreeAndNil(UndoHistory);
+  inherited;
+end;
+
+function TUndoSystem.UndoHistorySize: Integer;
+var
+  U: TUndoHistoryElement;
+begin
+  Result := 0;
+  for U in UndoHistory do
+    Result += U.Size;
+end;
+
+procedure TUndoSystem.RecordUndo(const UndoData: TUndoData; const SelectedComponent: TSelectedComponent; const ItemIndex: Integer; const TabIndex: Integer; const UndoComment: String);
+var
+  NewUndoElement: TUndoHistoryElement;
+  I: Integer;
+  NewUndoHistorySize: Integer;
+begin
+  ScheduleRecordUndoOnRelease := false;
+  if (UndoHistory.Count > 0) and (UndoData = UndoHistory[CurrentUndo].Data) then
+  begin
+    if (SelectedComponent = UndoHistory[CurrentUndo].Selected) or (UndoHistory[CurrentUndo].Selected = '') then
+      WriteLnLog('New Undo is identical to previous Undo record. Not saving.')
+    else
+    begin
+      UndoHistory[CurrentUndo].Selected := SelectedComponent;
+      WriteLnLog('New Undo is identical to previous Undo record. Only selection has changed from ' + UndoHistory[CurrentUndo].Selected + ' to ' + SelectedComponent + '. Not saving.');
+      OnUpdateUndo(Self);
+    end;
+    if (UndoComment <> '') and (UndoHistory[CurrentUndo].Comment = '') then
+    begin
+      WriteLnLog('The last undo comment was set to generic "Edit value", overwriting with better reason: ' + UndoComment);
+      UndoHistory[CurrentUndo].Comment := UndoComment;
+      OnUpdateUndo(Self);
+    end;
+    Exit;
+  end;
+  WriteLnLog('Saving Undo record. CurrentUndo = ' + IntToStr(CurrentUndo));
+  //Clean all next undo records if available;
+  for I := UndoHistory.Count - 1 downto CurrentUndo + 1 do
+    UndoHistory.Delete(I);
```

Comment:
This deletion can be done more easily and efficiently by just:

```
UndoHistory.Count := CurrentUndo;
```

Reply:
Done!